libraries/AP_Winch/AP_Winch_Servo.cpp

```diff
@@ -0,0 +1,95 @@
+#include <AP_Winch/AP_Winch_Servo.h>
+
+extern const AP_HAL::HAL& hal;
+
+void AP_Winch_Servo::init(const AP_WheelEncoder* wheel_encoder)
+{
+    _wheel_encoder = wheel_encoder;
+}
+
+void AP_Winch_Servo::update()
+{
+    // return immediately if no servo is assigned to control the winch
+    if (!SRV_Channels::function_assigned(SRV_Channel::k_winch)) {
+        return;
+    }
+
+    if (!_servo_out_init) {
+        // set servo output range
+        SRV_Channels::set_angle(SRV_Channel::k_winch,  1000);
+        _servo_out_init = true;
+    }
+
+    // return immediately if no wheel encoder
+    if (_wheel_encoder == nullptr) {
+        return;
+    }
+
+    // if not doing any control stop output pwm exit
+    if (config.state == AP_Winch::STATE_DISABLED) {
+        SRV_Channels::set_safety_limit(SRV_Channel::k_winch, SRV_Channel::SRV_CHANNEL_LIMIT_ZERO_PWM);
+        return;
+    }
+
+    // calculate dt since last iteration
+    uint32_t now = AP_HAL::millis();
+    float dt = (now - last_update_ms) / 1000.0f;
+    if (dt > 1.0f) {
+        dt = 0.0f;
+    }
+    last_update_ms = now;
+
+    // calculate latest rate
+    float distance = _wheel_encoder->get_distance(0);
+    float rate = 0.0f;
+    if (is_positive(dt)) {
+        rate = (distance - _last_distance) / dt;
+    }
+    _last_distance = distance;
+
+    // update distance from wheel encoder
+    config.length_curr = distance;
+
+    // if doing position control, calculate position error to desired rate
+    float rate_desired = 0.0f;
+    if (config.state == AP_Winch::STATE_POSITION) {
+        float position_error = config.length_desired - config.length_curr;
+        rate_desired = position_error * config.pos_p;
+    }
+
+    // if doing rate control, set desired rate
+    if (config.state == AP_Winch::STATE_RATE) {
+        rate_desired = config.rate_desired;
+    }
+
+    // calculate rate error and pass to pid controller
+    float rate_error = rate_desired - rate;
```

Comment:
Shouldn't rate desired always be clamped to be within the max winch rate? Otherwise a high power motor can actually throw the rope off the winch faster then the load can pull it down, and can lead to all sorts of tangling/wire problems below the vehicle.

Reply:
yes, it should be constrained, added!